tools/lib/addPipes.js

```diff
@@ -0,0 +1,308 @@
+const aStar = require('a-star');
+const jsgraphs = require('js-graph-algorithms');
+const Victor = require('victor');
+
+
+const PUMPJACK_EXIT_DIRECTION = {
+  0: { x: 2, y: -1 },
+  1: { x: 3, y: 0 },
+  2: { x: 0, y: 3 },
+  3: { x: -1, y: 2 }
+};
+
+const MAX_UNDERGROUND_REACH = 11; // Includes underground pipes
+
+const SIDES = [
+  { x: 0, y: -1 },
+  { x: 1, y: 0 },
+  { x: 0, y: 1 },
+  { x: -1, y: 0 },
+];
+
+const DIRECTION_FROM_OFFSET = {
+  '0,-1': 0,
+  '1,0': 2,
+  '0,1': 4,
+  '-1,0': 6
+};
+
+let ROTATE_ALL = false;
+let FLIP_ALL = false;
+
+function getPumpjackOutput(pumpjack, direction) {
+  const ROTATE_OFFSET = ((4 - ROTATE_ALL) % 4);
+  const ORDER = [0, -1, 2, 3];
+  if (typeof direction == 'undefined') direction = pumpjack.direction/2;
+  if (direction % 2 == 0) ORDER.reverse();
+  let offset = {
+    x: PUMPJACK_EXIT_DIRECTION[(direction + ROTATE_ALL) % 4].x,
+    y: PUMPJACK_EXIT_DIRECTION[(direction + ROTATE_ALL) % 4].y
+  };
+  offset.x = ORDER[(ORDER.indexOf(offset.x) + ROTATE_OFFSET) % ORDER.length];
+  offset.y = ORDER[(ORDER.indexOf(offset.y) + ROTATE_OFFSET) % ORDER.length];
+  if (FLIP_ALL) offset.x = 2 - offset.x;
+
+  return pumpjack.position.clone().add(offset);
+}
+
+function getStepFromDirection(direction) {
+  let out = new Victor(SIDES[direction].x, SIDES[direction].y);
+  if (FLIP_ALL) out.x = -out.x;
+  return out;
+}
+
+function connectPumpjack(bp, pumpjack, target, xLimit){
+  const hasBeacons = bp.entities.some(ent => ent.name == 'beacon');
+
+  // Bound the search space for improved speed
+  const boundsX = [bp.bottomLeft().x - 1, xLimit];
+  const boundsY = [bp.topRight().y - 1, bp.bottomLeft().y + 1];
+
+  const result = aStar({
+    start: [null, 'start'],
+
+    isEnd: ([node, meta]) => {
+      if (meta == 'start') return false;
+      if (node.x == target.x && node.y == target.y) return true;
+      let entity = bp.findEntity(node);
+      if (entity && entity.name == 'pipe') return true;
+      for (let i = 0; i < SIDES.length; i++) {
+        const entity = bp.findEntity(node.clone().add(SIDES[i]));
+        if (entity && entity.name == 'pipe') {
+          return true;
+        }
+      }
+      return false;
+    },
+
+    neighbor: ([node, meta]) => {
+      let isClear = (p) => (
+        p.x >= boundsX[0] && p.x <= boundsX[1] &&
+        p.y >= boundsY[0] && p.y <= boundsY[1] &&
+        (!bp.findEntity(p) || (bp.findEntity(p).name == 'beacon' ||
+                               bp.findEntity(p).name == 'pipe'))
+      );
+      let walkInDirection = (first, step) => {
+        let pos = first.clone();
+        let steps = [];
+        for (let j = 0; j < MAX_UNDERGROUND_REACH; j++){
+          if (!isClear(pos))
+            break;
+          steps.push(pos);
+          pos = pos.clone().add(step);
+        }
+        return steps;
+      };
+
+      let neighbors = [];
+      if (meta == 'start') {
+        for (let dir of [0, 1, 2, 3]) {
```

Comment:
I think it may be worth adding an option to allow the user to use the original rotation of the pumpjack (I believe just `pumpjack.direction/2` since some people may rather spend a bit more iron placing pipes over having to dig up every single pumpjack. Otherwise looks awesome :) Sorry for getting to this so late!

Reply:
Isn't this tool only really useful if you have construction robots and then you can just use a deconstruction planner?